File: backend/app/schemas/ai.py
"""AI Gateway request and response schemas."""
from typing import Any, Literal

from pydantic import BaseModel, Field


AIScenario = Literal[
    "project_draft",
    "plan_extract",
    "material_recommend",
    "material_copy",
    "plan_review",
    "campaign_diagnosis",
    "report_summary",
    "chat_general",
]


class AIMessage(BaseModel):
    role: Literal["system", "user", "assistant"]
    content: str = Field(..., max_length=8000)


class AIRunRequest(BaseModel):
    scenario: AIScenario
    project_id: str | None = None
    campaign_id: str | None = None
    material_id: str | None = None
    messages: list[AIMessage] = Field(default_factory=list)
    context: dict[str, Any] = Field(default_factory=dict)
    response_schema: dict[str, Any] | None = None


class AIUsagePayload(BaseModel):
    usage_log_id: str
    provider: str
    model: str
    input_tokens: int
    output_tokens: int
    total_tokens: int
    estimated_cost_usd: float
    daily_limit_remaining: int | None = None


class AIRunResponse(BaseModel):
    scenario: AIScenario
    status: Literal["draft", "suggested", "blocked", "failed"]
    output: dict[str, Any]
    usage: AIUsagePayload
    requires_human_confirm: bool = True


class AIUsageSummary(BaseModel):
    total_tokens: int
    estimated_cost_usd: float
    by_scenario: dict[str, Any]


class AIBudgetRequest(BaseModel):
    scope_type: Literal["user", "project"] = "user"
    scope_id: str | None = None
    daily_token_limit: int | None = Field(None, gt=0)
    monthly_token_limit: int | None = Field(None, gt=0)
    daily_cost_limit_usd: float | None = Field(None, ge=0)
    monthly_cost_limit_usd: float | None = Field(None, ge=0)
    enabled: bool = True

File: frontend/packages/main-app/src/api/ai.ts
import { http } from './http'

export type AIScenario =
  | 'project_draft'
  | 'plan_extract'
  | 'material_recommend'
  | 'material_copy'
  | 'plan_review'
  | 'campaign_diagnosis'
  | 'report_summary'
  | 'chat_general'

export interface AIMessage {
  role: 'system' | 'user' | 'assistant'
  content: string
}

export interface AIRunRequest {
  scenario: AIScenario
  project_id?: string
  campaign_id?: string
  material_id?: string
  messages?: AIMessage[]
  context?: Record<string, any>
  response_schema?: Record<string, any>
}

export interface AIRunResponse {
  scenario: AIScenario
  status: 'draft' | 'suggested' | 'blocked' | 'failed'
  output: Record<string, any>
  usage: {
    usage_log_id: string
    provider: string
    model: string
    input_tokens: number
    output_tokens: number
    total_tokens: number
    estimated_cost_usd: number
    daily_limit_remaining?: number
  }
  requires_human_confirm: boolean
}

export interface AIUsageLog {
  id: string
  project_id?: string
  campaign_id?: string
  scenario: AIScenario
  provider?: string
  model?: string
  input_tokens: number
  output_tokens: number
  total_tokens: number
  estimated_cost_usd: number
  status: string
  error_message?: string
  created_at?: string
}

export async function runAI(data: AIRunRequest): Promise<AIRunResponse> {
  return http.post<AIRunResponse>('/ai/run', data)
}

export async function getAIUsageSummary(params?: {
  project_id?: string
}): Promise<{
  total_tokens: number
  estimated_cost_usd: number
  by_scenario: Record<string, { total_tokens: number; estimated_cost_usd: number }>
}> {
  const query = new URLSearchParams()
  if (params?.project_id) query.set('project_id', params.project_id)
  const suffix = query.toString() ? `?${query.toString()}` : ''
  return http.get(`/ai/usage/summary${suffix}`)
}

export async function getAIUsageLogs(params?: {
  project_id?: string
  scenario?: AIScenario
  limit?: number
}): Promise<AIUsageLog[]> {
  const query = new URLSearchParams()
  if (params?.project_id) query.set('project_id', params.project_id)
  if (params?.scenario) query.set('scenario', params.scenario)
  if (params?.limit) query.set('limit', String(params.limit))
  const suffix = query.toString() ? `?${query.toString()}` : ''
  return http.get<AIUsageLog[]>(`/ai/usage/logs${suffix}`)
}

export async function setAIUsageBudget(data: {
  scope_type?: 'user' | 'project'
  scope_id?: string
  daily_token_limit?: number
  monthly_token_limit?: number
  daily_cost_limit_usd?: number
  monthly_cost_limit_usd?: number
  enabled?: boolean
}): Promise<Record<string, any>> {
  return http.post('/ai/usage/budget', data)
}

